### ISSUE TITLE
NPE at DNS lookup

### ISSUE THREAD

**Issue by emauro** [square/okhttp#3122]
I am not sure if it is my fault or not. I have very few reports of this error (very, very rare). See stack trace below:

java.lang.NullPointerException: Attempt to get length of null array
	at java.net.InetAddress.lookupHostByName(InetAddress.java:453)
	at java.net.InetAddress.getAllByNameImpl(InetAddress.java:252)
	at java.net.InetAddress.getAllByName(InetAddress.java:215)
	at okhttp3.Dns$1.lookup(Dns.java:39)
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:170)
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:136)
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:81)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:181)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:132)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:101)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.burgstaller.okhttp.AuthenticationCacheInterceptor.intercept(AuthenticationCacheInterceptor.java:40)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.abc.android.singleton.HTTPClient$1.intercept(HTTPClient.java:64)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:129)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
	at java.lang.Thread.run(Thread.java:818)

**Comment by vanshg** [comment]
Same error here. Version 3.4.1

**Comment by swankjesse** [comment]
Looks like a bug in the runtime. This needs to go to the OS maintainer to fix.

**Comment by vanshg** [comment]
Still seem to be getting this same error in my crash logs, but pretty rare. Not sure what's causing it, but it is occurring on version 3.6.0

**Comment by Larnoo** [comment]
Device: FRD-AL00
Android:  7.0

```
java.lang.NullPointerException: Attempt to get length of null array
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:108)
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
	at java.net.InetAddress.getAllByName(InetAddress.java:752)
	at okhttp3.Dns$1.lookup(Dns.java:39)
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:170)
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:136)
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:81)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:171)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:121)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:100)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:129)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
	at java.lang.Thread.run(Thread.java:776)
```


This is my crash log.
Android Platform Api 25 source code as follow.

```
private static InetAddress[] lookupHostByName(String host, int netId)
            throws UnknownHostException {
        BlockGuard.getThreadPolicy().onNetwork();
        // Do we have a result cached?
        Object cachedResult = addressCache.get(host, netId);
        if (cachedResult != null) {
            if (cachedResult instanceof InetAddress[]) {
                // A cached positive result.
                return (InetAddress[]) cachedResult;
            } else {
                // A cached negative result.
                throw new UnknownHostException((String) cachedResult);
            }
        }
        try {
            StructAddrinfo hints = new StructAddrinfo();
            hints.ai_flags = AI_ADDRCONFIG;
            hints.ai_family = AF_UNSPEC;
            // If we don't specify a socket type, every address will appear twice, once
            // for SOCK_STREAM and one for SOCK_DGRAM. Since we do not return the family
            // anyway, just pick one.
            hints.ai_socktype = SOCK_STREAM;
            InetAddress[] addresses = Libcore.os.android_getaddrinfo(host, hints, netId);
            // TODO: should getaddrinfo set the hostname of the InetAddresses it returns?
            for (InetAddress address : addresses) {
                address.holder().hostName = host;
            }
            addressCache.put(host, netId, addresses);
            return addresses;
        } catch (GaiException gaiException) {
            // If the failure appears to have been a lack of INTERNET permission, throw a clear
            // SecurityException to aid in debugging this common mistake.
            // http://code.google.com/p/android/issues/detail?id=15722
            if (gaiException.getCause() instanceof ErrnoException) {
                if (((ErrnoException) gaiException.getCause()).errno == EACCES) {
                    throw new SecurityException("Permission denied (missing INTERNET permission?)", gaiException);
                }
            }
            // Otherwise, throw an UnknownHostException.
            String detailMessage = "Unable to resolve host \"" + host + "\": " + Libcore.os.gai_strerror(gaiException.error);
            addressCache.putUnknownHost(host, netId, detailMessage);
            throw gaiException.rethrowAsUnknownHostException(detailMessage);
        }
    }
```

May be the line ` InetAddress[] addresses = Libcore.os.android_getaddrinfo(host, hints, netId);`
return a null array.

**Comment by forevervivi** [comment]
java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:108)
--
2 java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74)
3 java.net.InetAddress.getAllByName(InetAddress.java:752)
4 okhttp3.Dns$1.lookup(Dns.java:39)
5 okhttp3.internal.connection.RouteSelector.a(RouteSelector.java:171)
6 okhttp3.internal.connection.RouteSelector.d(RouteSelector.java:137)
7 okhttp3.internal.connection.RouteSelector.b(RouteSelector.java:82)
8 okhttp3.internal.connection.StreamAllocation.a(StreamAllocation.java:171)
9 okhttp3.internal.connection.StreamAllocation.a(StreamAllocation.java:121)
10 okhttp3.internal.connection.StreamAllocation.a(StreamAllocation.java:100)
11 okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
12 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:92)
13 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:67)
14 okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
15 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:92)
16 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:67)
17 okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
18 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:92)
19 okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
20 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:92)
21 okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.java:67)
22 okhttp3.RealCall.j(RealCall.java:185)
23 okhttp3.RealCall$AsyncCall.b(RealCall.java:135)
24 okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
25 java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
26 java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
27 java.lang.Thread.run(Thread.java:761)

![image](https://user-images.githubusercontent.com/3841809/29663754-48780a46-88ff-11e7-95e6-51ddb0794cf4.png)

both happened on android 7.0+.

**Comment by TheLester** [comment]
Same problem on Android 7 . How i can prevent it from crashing?

**Comment by Lee-swifter** [comment]
I've faced the same problem, crash will happen when you return null in `Dns.lookup()`, okhttp just call `List.isEmpty()` on it, so if you return null, you get NPE.

I think okhttp should use `if(list != null)` to ensure whether custom Dns returns null.

**Comment by jhass** [comment]
Can we have this reopened please?

**Comment by yschimke** [comment]
@swankjesse I can see the point, having an app crash because an app that *doesn't* customise Dns, but requiring a fix from the OS provider is frustrating.  Any objection to something like


```
  Dns SYSTEM = new Dns() {
    @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
      if (hostname == null) throw new UnknownHostException("hostname == null");
      try {
        return Arrays.asList(InetAddress.getAllByName(hostname));
      } catch (NullPointerException e) {
        throw new UnknownHostException("broken dns implementation", e);
      }
    }
  };
```

**Comment by swankjesse** [comment]
@yschinke yes, definitely want us to do that. Sorry I haven't done it yet!

**Comment by yschimke** [comment]
I can do it.

**Comment by yschimke** [comment]
https://github.com/square/okhttp/pull/3691

**Comment by swankjesse** [comment]
Released as OkHttp 3.9.1.
